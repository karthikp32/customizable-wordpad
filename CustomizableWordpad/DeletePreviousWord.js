import React from "react";
import { View, Text } from "react-native";



//deletes previous word from where cursor is located
//if cursor is in front of a space or other separator,
//then cursor moves to end of previous word
//else cursor stays at end of word

//To delete the previous word, the length of the previous word
//must be found by finding the amount of characters that must traveled through
//until the next space or separator is found
//then that characters equaling that amount will be deleted from the string
//that represents the text input by the user

const DeletePreviousWord = () => {

}

export default DeletePreviousWord;