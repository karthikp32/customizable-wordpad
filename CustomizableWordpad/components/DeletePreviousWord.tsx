import * as React from 'react';

import { Text, TextProps } from './Themed';

const separatorString = "\n\t\r,. "

export function DeletePreviousWord(currentIndex: number, sentence: string): string {
  let previousWordIndices = FindPreviousWordIndices(currentIndex, sentence);
  // return sentence.replace(previousWord, '');
  let previousWordSubstring = sentence.substring(previousWordIndices.startIndex, previousWordIndices.endIndex+1);
  return sentence.replace(previousWordSubstring, '');
}

export function FindPreviousWordIndices(currentIndex: number, sentence: string): any {
  let i = currentIndex;
  let indicesToDeleteWord = {startIndex: 0, endIndex: 0};
  if (!CheckIfCharacterIsASeparator(sentence.charAt(i))) {
    indicesToDeleteWord = FindIndicesOfPreviousWordOrSeparator(currentIndex, sentence);
  }
  else { 
    const separatorString = FindIndicesOfPreviousWordOrSeparator(currentIndex, sentence);
    let endIndexOfWord = separatorString.startIndex - 1;
    const word = FindIndicesOfPreviousWordOrSeparator(endIndexOfWord, sentence);
    indicesToDeleteWord.startIndex = word.startIndex;
    indicesToDeleteWord.endIndex = separatorString.endIndex;
  }
  return indicesToDeleteWord;
}

export function findEndIndexOfPreviousWord(word: string): number {
  return word.length - 1;
}

export function FindIndicesOfPreviousWordOrSeparator(currentIndex: number, sentence: string){
  let i = currentIndex;
  if (!CheckIfCharacterIsASeparator(sentence.charAt(i))) {
    while(i>=0 && !CheckIfCharacterIsASeparator(sentence.charAt(i-1))){
      i--;
    }
  }
  if (CheckIfCharacterIsASeparator(sentence.charAt(i))) {
    while(i>=0 && CheckIfCharacterIsASeparator(sentence.charAt(i-1))){
      i--;
    }
  }
  let indices = {startIndex: i, endIndex: currentIndex};
  return indices;
}

export function CheckIfCharacterIsASeparator(character: string): boolean {
  return separatorString.indexOf(character) > -1;
}