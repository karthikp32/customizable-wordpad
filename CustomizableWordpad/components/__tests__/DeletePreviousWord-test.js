import * as React from 'react';
import renderer from 'react-test-renderer';

import { DeletePreviousWord, CheckIfCharacterIsASeparator, FindIndicesOfPreviousWordOrSeparator, FindPreviousWordIndices } from '../DeletePreviousWord';

it(`given cursor is after a separator, deletePreviousWord() should delete the word 
before the separator`, () => {
    const sentence = 'Hello, my name is ';
    const currentIndex = 9;
    const expectedResult = 'Hello, name is ';
    const actualResult = DeletePreviousWord(currentIndex, sentence);
    expect(actualResult).toBe(expectedResult);
});

it(`given cursor is after a word, deletePreviousWord() should delete the previous
word`, () => {
    const sentence = 'Hello, my name is ';
    const currentIndex = 8;
    const expectedResult = 'Hello,  name is ';
    const actualResult = DeletePreviousWord(currentIndex, sentence);
    expect(actualResult).toBe(expectedResult);
});

it(`given cursor is after a separator, findPreviousWord() should find the indices
of the word and separators`, () => {
    const sentence = 'Hello, my name is ';
    const currentIndex = 9;
    const expectedIndices = {startIndex: 7, endIndex: 9};
    const actualIndices = FindPreviousWordIndices(currentIndex, sentence);
    expect(actualIndices).toStrictEqual(expectedIndices);
});

it(`given cursor is after a word, findPreviousWord() should find the indices
of the word`, () => {
    const sentence = 'Hello, my name is ';
    const currentIndex = 8;
    const expectedIndices = {startIndex: 7, endIndex: 8};
    const actualIndices = FindPreviousWordIndices(currentIndex, sentence);
    expect(actualIndices).toStrictEqual(expectedIndices);
});


it(`given cursor is after a word, deletePreviousWord() should delete the previous
word`, () => {
  const inputString = 'Hello, my name is';
  const expectedString = 'Hello, my name';
  const actualString = 'Hello, my name';
  expect(actualString).toBe(expectedString);
});


it(`given cursor is after a separator, FindIndicesOfPreviousWordOrSeparator() should find the indices of the previous word
or separator`, () => {
  const sentence = 'Hello, my name is ';
  const currentIndex = 6;
  const expectedIndices = {startIndex: 5, endIndex: 6};
  const actualIndices = FindIndicesOfPreviousWordOrSeparator(currentIndex, sentence);
  expect(actualIndices).toStrictEqual(expectedIndices);
});

it(`given cursor is after a word, FindIndicesOfPreviousWordOrSeparator() should find the indices of the previous word
or separator`, () => {
  const sentence = 'Hello, my name is ';
  const currentIndex = 4;
  const expectedIndices = {startIndex: 0, endIndex: 4};
  const actualIndices = FindIndicesOfPreviousWordOrSeparator(currentIndex, sentence);
  expect(actualIndices).toStrictEqual(expectedIndices);
});

it(`given character is a separator, checkIfCharacterIsASeparator() returns true`, () => {
  const inputCharacter = ' ';
  const actualAnswer = CheckIfCharacterIsASeparator(inputCharacter);
  expect(actualAnswer).toBeTruthy();
});

it(`given character is a letter, checkIfCharacterIsASeparator() returns true`, () => {
    const inputCharacter = 'a';
    const actualAnswer = CheckIfCharacterIsASeparator(inputCharacter);
    expect(actualAnswer).toBeFalsy();
});


it(`given character is a number, checkIfCharacterIsASeparator() returns true`, () => {
    const inputCharacter = '2';
    const actualAnswer = CheckIfCharacterIsASeparator(inputCharacter);
    expect(actualAnswer).toBeFalsy();
});  


