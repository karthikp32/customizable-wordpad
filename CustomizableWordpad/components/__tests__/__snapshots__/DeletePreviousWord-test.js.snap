// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given cursor is after a separator, deletePreviousWord() should delete the word 
before the separator 1`] = `"Hello, my name is "`;
